### PR TITLE
SmartCalcs Connect

### PR DESCRIPTION
**Pull request by dallendalton**
Currently when setting up the plugin the user is required to navigate to the TaxJar app, get their API token, and then return to the plugin setting page to save the token and link their store with TaxJar. This PR provides a UI that allows the customer to simple click a connect button during setup, which will open a login to TaxJar, after signing in their API token will be automatically added and saved.

**Click-Test Versions**

- [X] Woo 4.0
- [X] Woo 3.9
- [X] Woo 3.8
- [X] Woo 3.7
- [X] Woo 3.6
- [X] Woo 3.5
- [X] Woo 3.4
- [X] Woo 3.3
- [X] Woo 3.2
- [X] Woo 3.1
- [X] Woo 3.0

**Specs Passing**

- [X] Woo 4.0
- [X] Woo 3.9
- [X] Woo 3.8
- [X] Woo 3.7
- [X] Woo 3.6
- [X] Woo 3.5
- [X] Woo 3.4
- [X] Woo 3.3
- [X] Woo 3.2
- [X] Woo 3.1
- [X] Woo 3.0

### CODE REVIEW

File: includes/class-wc-taxjar-connection.php [hunk 1]
```diff
@@ -61,7 +61,7 @@ private function check_status() {
 			if ( $this->integration->post_or_setting( 'api_token' ) && isset( $body->valid ) && false === $body->valid ) {
 				$description .= '<span style="color: #ff0000;"><strong>';
 				$description .= 'It looks like your API token is invalid.';
-				$description .= sprintf( '<br><a href="%s" target="_blank">Please review your API token.</a>', $this->integration->app_uri . 'account#api-access' );
+				$description .= sprintf( '<br><span style="color: black;">Please attempt to reconnect to TaxJar or </span><a href="%s" target="_blank">review your API token here.</a>', $this->integration->app_uri . 'account#api-access' );
```
Comment (codymullins) [inline includes/class-wc-taxjar-connection.php:64]: I think we could remove the `here` wording. I think the `review your API Token` terminology is sufficient. "Click here" type wording in links is generally discouraged.

File: includes/class-wc-taxjar-integration.php [hunk 2]
```diff
@@ -269,26 +271,70 @@ public function get_settings( $current_section = '' ) {
...
 			);
 
+			if ( $api_token_valid ) {
+				$connected_email = $this->post_or_setting( 'connected_email' );
+
+				if ( $connected_email ) {
+					array_push( $settings, array(
+						'title' => '',
+						'type'  => 'title',
+						'desc'  => '<div class="taxjar-connected><span class="dashicons dashicons-yes-alt"></span><p>' . $connected_email . '</p></div>',
```
Comment (codymullins) [inline includes/class-wc-taxjar-integration.php:284]: Looks like you're missing a quotation.

```
class="taxjar-connected>
```

**should be**

```
class="taxjar-connected">
```

File: includes/class-wc-taxjar-integration.php [hunk 3]
```diff
@@ -269,26 +271,70 @@ public function get_settings( $current_section = '' ) {
...
+						'id'	=> 'connected-to-taxjar'
+					) );
+				} else {
+					array_push( $settings, array(
+						'title' => '',
+						'type'  => 'title',
+						'desc'  => '<div class="taxjar-connected><span class="dashicons dashicons-yes-alt"></span><p>Connected To TaxJar</p></div>',
```
Comment (codymullins) [inline includes/class-wc-taxjar-integration.php:291]: Missing quotation here as well.

File: includes/class-wc-taxjar-integration.php [hunk 4]
```diff
@@ -269,26 +271,70 @@ public function get_settings( $current_section = '' ) {
...
+						'id'	=> 'connected-to-taxjar'
+					) );
+				}
+
+				array_push( $settings, array(
+					'title' => '',
+					'type'  => 'title',
+					'desc'  => __( '<button id="disconnect-from-taxjar" name="disconnect-from-taxjar" class="button-primary" type="submit" value="Disconnect">Disconnect From TaxJar</button><p><a href="#" id="connect-manual-edit">Click here to edit API Token manually.</a></p>', 'wc-taxjar' ),
```
Comment (codymullins) [inline includes/class-wc-taxjar-integration.php:299]: I would rephrase this link text to avoid "click here" type wording.

Maybe just `Edit API Token` would be sufficient?

File: includes/js/wc-taxjar-admin.js [hunk 5]
```diff
@@ -8,6 +8,73 @@ jQuery( document ).ready( function() {
...
+			}
+		};
+
+		var disconnect_taxjar = function( e ) {
+			$( '#woocommerce_taxjar-integration_settings\\[api_token\\]' ).val( '' )
+			$( '#woocommerce_taxjar-integration_settings\\[connected_email\\]' ).val( '' );
+
+			var input = $( "<input>" ).attr( "type", "hidden" ).attr( "name", "save" ).val( "Save changes" );
+			$( '#mainform' ).append( input );
+			//$( '#mainform button.woocommerce-save-button' ).click();
```
Comment (codymullins) [inline includes/js/wc-taxjar-admin.js:30]: Can you remove this commented code?

File: includes/js/wc-taxjar-admin.js [hunk 6]
```diff
@@ -8,6 +8,73 @@ jQuery( document ).ready( function() {
...
+			var dualScreenLeft = window.screenLeft != undefined ? window.screenLeft : screen.left;
+			var dualScreenTop = window.screenTop != undefined ? window.screenTop : screen.top;
+			var width = window.innerWidth ? window.innerWidth : document.documentElement.clientWidth ? document.documentElement.clientWidth : screen.width;
+			var height = window.innerHeight ? window.innerHeight : document.documentElement.clientHeight ? document.documentElement.clientHeight : screen.height;
+			var left = ((width / 2) - (w / 2)) + dualScreenLeft;
+			var top = ((height / 2) - (h / 2)) + dualScreenTop;
+
+			window.popup = window.open(url, title, 'scrollbars=yes, width=' + w + ', height=' + h + ', top=' + top + ', left=' + left);
+
+			if (window.focus) window.popup.focus();
```
Comment (codymullins) [inline includes/js/wc-taxjar-admin.js:77]: I would recommend always using braces in `if` statements :)

File: includes/class-wc-taxjar-integration.php [hunk 7]
```diff
@@ -329,7 +375,7 @@ public function get_settings( $current_section = '' ) {
 						'type' => 'title',
 					);
 					$settings[] = array(
-						'title'    => __( 'Ship From Address', 'wc-taxjar' ),
+						'title'	=> __( 'Ship From Address', 'wc-taxjar' ),
```
Comment (fastdivision) [inline includes/class-wc-taxjar-integration.php:378]: @dallendalton Was this intended?

File: includes/class-wc-taxjar-integration.php [hunk 8]
```diff
@@ -1557,6 +1605,17 @@ public function load_taxjar_admin_assets() {
 		wp_enqueue_style( 'jquery-ui-datepicker' );
 	}
 
+	 /**
+	  * Generates TaxJar connect popup url
+	  *
+	  * @return string - TaxJar connect popup url
+	  */
+	public function get_connect_url() {
+		$connect_url =  $this->app_uri . 'smartcalcs/connect/woo/?store=' . urlencode( get_bloginfo( 'url' ) );
```
Comment (fastdivision) [inline includes/class-wc-taxjar-integration.php:1614]: @dallendalton Extra space between `=` and `$this`

File: includes/class-wc-taxjar-integration.php (hunk 7 above)
Comment (dallendalton) [reply includes/class-wc-taxjar-integration.php]: No it was not, it is resolved in the latest commit.
